test/framework/deployments/postgres/universal.go

```diff
@@ -0,0 +1,141 @@
+package postgres
+
+import (
+	"bytes"
+	"os/exec"
+	"regexp"
+
+	"github.com/gruntwork-io/terratest/modules/retry"
+	"github.com/gruntwork-io/terratest/modules/testing"
+	"github.com/pkg/errors"
+
+	"github.com/kumahq/kuma/test/framework"
+	"github.com/kumahq/kuma/test/framework/deployments"
+)
+
+const (
+	AppKumaCP                = framework.AppKumaCP
+	AppPostgres              = framework.AppPostgres
+	StoreTypePostgres        = framework.StoreTypePostgres
+	PostgresImage            = framework.PostgresImage
+	PostgresEnvVarUser       = framework.PostgresEnvVarUser
+	PostgresEnvVarPassword   = framework.PostgresEnvVarPassword
+	PostgresEnvVarDB         = framework.PostgresEnvVarDB
+	EnvStoreType             = framework.EnvStoreType
+	EnvStorePostgresUser     = framework.EnvStorePostgresUser
+	EnvStorePostgresPassword = framework.EnvStorePostgresPassword
+	EnvStorePostgresDBName   = framework.EnvStorePostgresDBName
+	EnvStorePostgresPort     = framework.EnvStorePostgresPort
+	EnvStorePostgresHost     = framework.EnvStorePostgresHost
+	DefaultPostgresPort      = framework.DefaultPostgresPort
+	DefaultPostgresUser      = framework.DefaultPostgresUser
+	DefaultPostgresPassword  = framework.DefaultPostgresPassword
+	DefaultPostgresDBName    = framework.DefaultPostgresDBName
+	AfterCreateMainApp       = framework.AfterCreateMainApp
+)
+
+type universalDeployment struct {
+	*deployments.DockerDeployment
+}
+
+var _ Deployment = &universalDeployment{}
+
+func (u *universalDeployment) Name() string {
+	return AppPostgres
+}
+
+func (u *universalDeployment) Deploy(cluster framework.Cluster) error {
+	if err := u.AllocatePublicPortsFor("5432"); err != nil {
+		return err
+	}
+
+	err := u.
+		WithName(cluster.Name()+"_"+u.Name()).
+		WithTestingT(cluster.GetTesting()).
+		WithNetwork("kind").
+		WithImage(PostgresImage).
+		WithEnvVar(PostgresEnvVarUser, DefaultPostgresUser).
+		WithEnvVar(PostgresEnvVarPassword, DefaultPostgresPassword).
+		WithEnvVar(PostgresEnvVarDB, DefaultPostgresDBName).
+		RunContainer()
+	if err != nil {
+		return err
+	}
+
+	ip, err := u.GetIP()
+	if err != nil {
+		return err
+	}
+
+	cluster.
+		WithEnvVar(AppKumaCP, EnvStoreType, string(StoreTypePostgres)).
+		WithEnvVar(AppKumaCP, EnvStorePostgresUser, DefaultPostgresUser).
+		WithEnvVar(AppKumaCP, EnvStorePostgresPassword, DefaultPostgresPassword).
+		WithEnvVar(AppKumaCP, EnvStorePostgresDBName, DefaultPostgresDBName).
+		WithEnvVar(AppKumaCP, EnvStorePostgresPort, DefaultPostgresPort).
+		WithEnvVar(AppKumaCP, EnvStorePostgresHost, ip).
+		WithHookFn(AppKumaCP, AfterCreateMainApp, RunDBMigration)
```

Comment:
I hate to ask for more changes, but I think we are mixing a couple of things here.
The idea of deployments is that they are independent unit of deployments that you place on the cluster. I don't think Postgres deployment should know anything about Kuma CP (envs, hooks)

I think the flow should be that Postgres deployment exposes bunch of settings that Kuma CP can then use. For example

```
err = NewClusterSetup().
        Install(postgres.Install()).
        Setup(universal)
Expect(err).ToNot(HaveOccurred())

err = NewClusterSetup().
        Install(Kuma(core.Standalone, WithPostgres(postgres.From(universal).Settings()))).
        Setup(universal)
Expect(err).ToNot(HaveOccurred())
```

Reply:
Addressed